client/rpc.go

```diff
@@ -46,19 +47,53 @@ func (c *Client) StreamingRpcHandler(method string) (structs.StreamingRpcHandler
 	return c.streamingRpcs.GetHandler(method)
 }
 
+// Given a type that is or eventually points to a concrete type with an embedded QueryOptions
```

Comment:
These reflection-based functions are really here because we've removed the `HasTimeout` from the `RPCInfo` interface which was implemented by both the `QueryOptions` and the `WriteRequest`. It's been replaced only on the `QueryOptions` side by `TimeToBlock`. But it looks to me like all cases where we call `TimeToBlock`, we are testing for a non-0 result.

So couldn't we avoid this by having `TimeToBlock` on the `RPCInfo` interface, having `WriteRequest` implement it, but have `WriteRequest` always return 0? That seems cheaper to me if we can make it work.

Reply:
I originally thought about this but for some reason thought it was better to use reflection, maybe it was early in my thinking before it evolved a bit. Either way now I don't see any reason why not so I updated the PR.